File: harvester/src/main/java/harvester/PageLoader.java
/**
 * 
 */
package harvester;

import java.io.IOException;
import java.io.InputStream;
import java.net.URL;
import java.net.URLConnection;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import org.apache.commons.io.IOUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.google.gson.JsonObject;

/**
 * @author mattias
 * 
 */
public class PageLoader {
	private final static int NUMBER_OF_PAGES_TO_FETCH = 10;

	private final static Logger logger = LoggerFactory
			.getLogger(PageLoader.class);

	private SearchProvider searchProvider;

	private final PageLoaderReporter reporter;

	public PageLoader() {
		this(PageLoaderReporter.VOID_REPORTER);
	}

	public PageLoader(PageLoaderReporter reporter) {
		this.reporter = reporter;
	}

	public Iterable<HtmlPageData> getPagesForKeyword(String keyword)
			throws HarvesterException {
		List<String> urls = collectUrlsFromSearchResult(keyword);
		return loadPagesFromUrls(urls);
	}

	private List<String> collectUrlsFromSearchResult(String keyword) {
		List<String> collectedUrls = new ArrayList<String>();
		for (int pageStart = 1; collectedUrls.size() < NUMBER_OF_PAGES_TO_FETCH
				&& (pageStart > 0);) {

			logger.info("Searching for keyword {}, starting at {}", keyword,
					pageStart);
			JsonObject searchResult = searchProvider
					.searchForKeywordStartingAtSpecifiedIndex(keyword,
							pageStart);

			collectedUrls.addAll(Arrays
					.asList(GoogleSearchResultExtractorUtility
							.findUrls(searchResult)));

			pageStart = GoogleSearchResultExtractorUtility
					.extractNextStartIndex(searchResult);
		}
		return collectedUrls.subList(0,
				Math.min(collectedUrls.size(), NUMBER_OF_PAGES_TO_FETCH));
	}

	private Iterable<HtmlPageData> loadPagesFromUrls(List<String> urls) {
		List<HtmlPageData> pages = new ArrayList<HtmlPageData>();
		for (String url : urls) {
			HtmlPageData page = loadPageFromUrl(url);
			if (page != null) {
				pages.add(page);
			}
		}

		return pages;
	}

	private HtmlPageData loadPageFromUrl(String urlString) {
		InputStream in = null;
		try {
			in = createPageInputStream(urlString);
			if (in == null) {
				return null;
			}

			logger.info("Loading page {}", urlString);
			reporter.progress(String.format("Loading data from %s", urlString));

			return createHtmlFromPageContent(IOUtils.toString(in));
		} catch (IOException e) {
			logger.error("Could not load page for url " + urlString, e);
			reporter.progress(String.format(
					"Could not load data from %s, skipping (reason=%s)",
					urlString, e.getMessage()));
			return null;
		} finally {
			IOUtils.closeQuietly(in);
		}
	}

	private HtmlPageData createHtmlFromPageContent(String pageAsString) {
		HtmlElementFinder finder = new HtmlElementFinder(pageAsString);

		return new HtmlPageData(finder.getTitle(), finder.getMetaValue(),
				finder.getBodyValues());
	}

	protected InputStream createPageInputStream(String urlString)
			throws IOException {
		URL url = new URL(urlString);
		URLConnection conn = url.openConnection();
		InputStream in = conn.getInputStream();

		return in;
	}

	public void setSearchProvider(SearchProvider sp) {
		this.searchProvider = sp;
	}
}

File: harvester-scala/src/main/scala/harvester/PageLoader.scala
package harvester

import java.io.IOException
import java.io.InputStream
import java.io.UnsupportedEncodingException
import java.net.MalformedURLException
import java.net.URL
import java.net.URLConnection
import java.net.URLEncoder

import scala.collection.JavaConversions._
import scala.util.control.Exception._

import com.google.gson._

import SearchProvider._
import harvester.ElementQualifier._
import net.htmlparser.jericho.CharacterReference
import net.htmlparser.jericho.HTMLElementName
import net.htmlparser.jericho.MasonTagTypes
import net.htmlparser.jericho.MicrosoftConditionalCommentTagTypes
import net.htmlparser.jericho.PHPTagTypes
import net.htmlparser.jericho.Source
import net.htmlparser.jericho.TextExtractor

private object PageLoader {
  val NUMBER_OF_PAGES_TO_FETCH = 10;
}

private object HtmlElementFinder {
  MicrosoftConditionalCommentTagTypes.register();
  PHPTagTypes.register();
  PHPTagTypes.PHP_SHORT.deregister(); // remove PHP short tags for this
  // example otherwise they override
  // processing instructions
  MasonTagTypes.register();
}
private class HtmlElementFinder(private val page: String) {

  val source = new Source(page);
  source.fullSequentialParse();

  val title = getTitle(source)
  val keywords = getMetaValue(source, "keywords")
  val bodyContent = searchBody(source)

  private def searchBody(source: Source) = {
    var body = Map[ElementQualifier, String]()
    for (htmlTagName <- ElementQualifier.values) {

      source.getAllElements(htmlTagName.toString().toLowerCase()).foreach(hTag => {
        val content = new TextExtractor(hTag).toString();
        body.get(htmlTagName) match {
          case None => body += htmlTagName -> content
          case Some(value) => body += htmlTagName -> (value + " " + content)
        }
      })
    }
    body
  }

  def getBodyValues(q: ElementQualifier) = {
    bodyContent.get(q);
  }

  private def getTitle(source: Source) = {
    val titleElement = source.getFirstElement(HTMLElementName.TITLE)
    if (titleElement == null) {
      "";
    }

    // TITLE element never contains other tags so just decode it collapsing
    // whitespace:
    CharacterReference.decodeCollapseWhiteSpace(titleElement.getContent());
  }

  private def getMetaValue(source: Source, key: String) = {
    var pos = 0
    while (pos < source.length()) {
      val startTag = source.getNextStartTag(pos, "name", key, false);
      if (startTag == null)
        null;
      if (startTag.getName().equals(HTMLElementName.META)) {
        // Attribute values are automatically decoded
        startTag.getAttributeValue("content");
      }
      pos = startTag.getEnd();
    }
    "";
  }
}

private object SearchProvider {
  val SEARCH_URL_PATTERN = "https://www.googleapis.com/customsearch/v1?" + "cx=008238868649727884978:msi2xt9aatg&key=AIzaSyDqC8D822At_Sj2LumIu5a4au1J0gPEVB8&nu" + "&q=%s" + "&start=%d"
}

class SearchProvider {

  def searchForKeywordStartingAtSpecifiedIndex(keyword: String, startIndex: Int) = {
    catching(classOf[MalformedURLException], classOf[UnsupportedEncodingException]) opt {
      loadDataFromURL(format(SEARCH_URL_PATTERN, URLEncoder.encode(keyword, "utf-8"), startIndex));
    }
  }

  def loadDataFromURL(url: String) = new JsonParser().parse(scala.io.Source.fromURL(url).bufferedReader()).asInstanceOf[JsonObject]
}

object GoogleSearchResultExtractorUtility {
  def findUrls(input: JsonObject) = {
    input.get("items").asInstanceOf[JsonArray].filter(elem => elem.isJsonObject() && elem.isInstanceOf[JsonObject]).map(_.asInstanceOf[JsonObject].get("link").getAsString).toList
  }

  def extractNextStartIndex(searchResult: JsonObject) = {
    val nextPageElement = searchResult.get("queries").asInstanceOf[JsonObject].get("nextPage")
    if (nextPageElement == null || nextPageElement.isJsonNull()) {
      -1
    } else {
      nextPageElement.asInstanceOf[JsonArray].get(0).asInstanceOf[JsonObject].get("startIndex").getAsInt
    }
  }
}

object VoidPageLoaderReporter extends PageLoaderReporter

trait PageLoaderReporter {
  def report(message: String): Unit = println("Reportin " + message)
}

class PageLoader(reporter: PageLoaderReporter) {
  def this() = this(VoidPageLoaderReporter)

  var searchProvider: SearchProvider = null

  def getPagesForKeyword(keyword: String): Seq[HtmlPageData] = collectUrlsFromSearchResult(keyword).map(loadPageFromUrl).flatten;

  private def collectUrlsFromSearchResult(keyword: String) = {
    def collectUrlsFromSearchResult(keyword: String, start: Int, result: List[String]): List[String] = {
      if (start < 0 || result.size > 10) {
        result
      }

      searchProvider.searchForKeywordStartingAtSpecifiedIndex(keyword, start) match {
        case None => result
        case Some(searchResult) => collectUrlsFromSearchResult(keyword, GoogleSearchResultExtractorUtility.extractNextStartIndex(searchResult), result ::: GoogleSearchResultExtractorUtility.findUrls(searchResult))
      }
    }

    collectUrlsFromSearchResult(keyword, 1, List[String]())
  }

  private def loadPageFromUrl(urlString: String) = {
    catching(classOf[IOException]) opt {
      val finder = new HtmlElementFinder(scala.io.Source.fromURL(urlString).mkString);
      new HtmlPageData(finder.title, finder.keywords, finder.bodyContent)
    }
  }

}